File: packages/phovoice-engine/core/vietnamese_itn.py
"""Vietnamese Inverse Text Normalization (ITN) Module.
Converts spoken Vietnamese numbers, dates, times, currencies, and percentages into digits.
Examples:
  - "năm mươi lăm phần trăm" -> "55%"
  - "hai mươi ba" -> "23"
  - "một triệu hai trăm nghìn đồng" -> "1.200.000 đồng"
  - "ngày mười lăm tháng tám năm hai nghìn không trăm hai mươi tư" -> "ngày 15 tháng 8 năm 2024"
  - "không chín không năm một hai ba bốn năm sáu" -> "0905123456"
"""
from __future__ import annotations

import re
from typing import List, Tuple

# Mapping of basic units
UNITS = {
    "không": 0, "linh": 0, "lẻ": 0,
    "một": 1, "mốt": 1,
    "hai": 2,
    "ba": 3,
    "bốn": 4, "tư": 4,
    "năm": 5, "lăm": 5,
    "sáu": 6,
    "bảy": 7, "bẩy": 7,
    "tám": 8,
    "chín": 9,
}

DIGIT_WORDS = {
    "không": "0", "một": "1", "hai": "2", "ba": "3", "bốn": "4", "tư": "4",
    "năm": "5", "sáu": "6", "bảy": "7", "bẩy": "7", "tám": "8", "chín": "9"
}

def parse_vietnamese_number_words(words: List[str]) -> int | None:
    """Parse a list of Vietnamese number words up to billions into an integer."""
    if not words:
        return None
    
    total = 0
    current_billions = 0
    current_millions = 0
    current_thousands = 0
    current_hundreds = 0
    current_val = 0
    
    i = 0
    while i < len(words):
        w = words[i].lower()
        
        if w == "tỷ" or w == "ti":
            chunk = (current_val + current_hundreds + current_thousands + current_millions) or 1
            current_billions += chunk * 1_000_000_000
            current_millions = current_thousands = current_hundreds = current_val = 0
        elif w == "triệu":
            chunk = (current_val + current_hundreds + current_thousands) or 1
            current_millions += chunk * 1_000_000
            current_thousands = current_hundreds = current_val = 0
        elif w in ("nghìn", "ngàn"):
            chunk = (current_val + current_hundreds) or 1
            current_thousands += chunk * 1_000
            current_hundreds = current_val = 0
        elif w == "trăm":
            chunk = current_val or 1
            current_hundreds = chunk * 100
            current_val = 0
        elif w in ("mươi", "chục"):
            chunk = current_val or 1
            current_val = chunk * 10
        elif w == "mười":
            current_val += 10
        elif w in ("linh", "lẻ"):
            pass
        elif w in UNITS:
            current_val += UNITS[w]
        else:
            return None
        i += 1
        
    total = current_billions + current_millions + current_thousands + current_hundreds + current_val
    return total

def format_number_with_dots(n: int) -> str:
    """Format large numbers with dot separators e.g. 1200000 -> 1.200.000"""
    if n >= 10000:
        return f"{n:,}".replace(",", ".")
    return str(n)


class VietnameseITN:
    """Inverse Text Normalization for Vietnamese speech recognition outputs."""

    # Words that form numbers
    NUM_WORDS = set(UNITS.keys()) | {"mươi", "mười", "chục", "trăm", "nghìn", "ngàn", "triệu", "tỷ", "ti"}

    @classmethod
    def normalize_phone_numbers(cls, text: str) -> str:
        """Convert digit sequences like 'không chín không năm một hai ba bốn năm sáu' -> '0905123456'."""
        tokens = text.split()
        out = []
        i = 0
        while i < len(tokens):
            # Check if this token starts a sequence of 4 or more isolated digit words
            j = i
            digits = []
            while j < len(tokens) and tokens[j].lower() in DIGIT_WORDS:
                digits.append(DIGIT_WORDS[tokens[j].lower()])
                j += 1
            
            # If 4 or more consecutive isolated single digits, convert to phone/serial number
            if len(digits) >= 4:
                out.append("".join(digits))
                i = j
            else:
                out.append(tokens[i])
                i += 1
        return " ".join(out)

    @classmethod
    def normalize_percentages(cls, text: str) -> str:
        """Convert 'X phần trăm' -> 'X%'."""
        def repl(match):
            num_str = match.group(1).strip()
            # Convert num_str to digits if words
            words = num_str.split()
            val = parse_vietnamese_number_words(words)
            if val is not None:
                return f"{val}%"
            return f"{num_str}%"

        # Regex for [number words] phần trăm
        pattern = r'\b((?:(?:không|một|mốt|hai|ba|bốn|tư|năm|lăm|sáu|bảy|bẩy|tám|chín|mười|mươi|chục|trăm|linh|lẻ)\s+)+)phần trăm\b'
        text = re.sub(pattern, repl, text, flags=re.IGNORECASE)
        text = re.sub(r'(\d+)\s+phần trăm\b', r'\1%', text, flags=re.IGNORECASE)
        return text

    @classmethod
    def normalize_dates_and_times(cls, text: str) -> str:
        """Convert 'ngày mười lăm tháng tám năm hai nghìn không trăm hai mươi tư' -> 'ngày 15 tháng 8 năm 2024'."""
        # Ngày X
        def repl_day(m):
            words = m.group(1).split()
            v = parse_vietnamese_number_words(words)
            return f"ngày {v}" if v is not None and 1 <= v <= 31 else m.group(0)
        
        text = re.sub(
            r'\bngày\s+((?:không|mồng|mùng|một|mốt|hai|ba|bốn|tư|năm|lăm|sáu|bảy|bẩy|tám|chín|mười|mươi|chục|hai mươi|ba mươi)(?:\s+(?:một|mốt|hai|ba|bốn|tư|năm|lăm|sáu|bảy|bẩy|tám|chín))?)\b',
            repl_day,
            text,
            flags=re.IGNORECASE
        )

        # Tháng Y
        def repl_month(m):
            words = m.group(1).split()
            v = parse_vietnamese_number_words(words)
            return f"tháng {v}" if v is not None and 1 <= v <= 12 else m.group(0)

        text = re.sub(
            r'\btháng\s+((?:một|hai|ba|bốn|tư|năm|sáu|bảy|bẩy|tám|chín|mười|mười một|mười hai))\b',
            repl_month,
            text,
            flags=re.IGNORECASE
        )

        # Năm Z
        def repl_year(m):
            words = m.group(1).split()
            v = parse_vietnamese_number_words(words)
            return f"năm {v}" if v is not None and 1900 <= v <= 2100 else m.group(0)

        text = re.sub(
            r'\bnăm\s+((?:(?:một|hai)\s+nghìn(?:\s+(?:không|lẻ|linh|một|hai|ba|bốn|năm|sáu|bảy|tám|chín|mười|mươi|trăm))*))\b',
            repl_year,
            text,
            flags=re.IGNORECASE
        )

        # Giờ / Phút
        def repl_time(m):
            h_words = m.group(1).split()
            h_val = parse_vietnamese_number_words(h_words)
            p_words = m.group(2).split() if m.group(2) else None
            p_val = parse_vietnamese_number_words(p_words) if p_words else None
            
            if h_val is not None:
                if p_val is not None:
                    return f"{h_val}h{p_val:02d}"
                return f"{h_val}h"
            return m.group(0)

        text = re.sub(
            r'\b((?:không|một|hai|ba|bốn|tư|năm|sáu|bảy|bẩy|tám|chín|mười|mười một|mười hai|mười ba|mười bốn|mười lăm|mười sáu|mười bảy|mười tám|mười chín|hai mươi|hai mươi mốt|hai mươi hai|hai mươi ba|hai mươi bốn))\s+giờ(?:\s+((?:(?:không|lẻ|linh|một|hai|ba|bốn|tư|năm|lăm|sáu|bảy|tám|chín|mười|mươi)\s*)+)(?:phút)?)?\b',
            repl_time,
            text,
            flags=re.IGNORECASE
        )

        return text

    @classmethod
    def normalize_general_numbers(cls, text: str) -> str:
        """Find spans of Vietnamese number words and convert them to digits."""
        words = text.split()
        if not words:
            return text

        result = []
        i = 0
        while i < len(words):
            # Check if words[i] is a number word
            clean_w = re.sub(r'[^\w\s]', '', words[i].lower())
            
            if clean_w in cls.NUM_WORDS:
                # Accumulate consecutive number words
                num_span = [clean_w]
                original_span = [words[i]]
                j = i + 1
                while j < len(words):
                    next_clean = re.sub(r'[^\w\s]', '', words[j].lower())
                    if next_clean in cls.NUM_WORDS:
                        num_span.append(next_clean)
                        original_span.append(words[j])
                        j += 1
                    else:
                        break
                
                # Try parsing the accumulated number span
                # We only convert if:
                # 1. Length of span >= 2 (e.g. "hai mươi", "một trăm", "mười lăm")
                # 2. Or single word is a compound magnitude (e.g. "nghìn", "triệu")
                val = parse_vietnamese_number_words(num_span) if len(num_span) >= 2 else None
                
                if val is not None and val > 0:
                    formatted = format_number_with_dots(val)
                    # Check if the last word had punctuation attached (e.g. "hai mươi,")
                    last_orig = original_span[-1]
                    punct_suffix = re.search(r'([,.:;?!]+)$', last_orig)
                    if punct_suffix:
                        formatted += punct_suffix.group(1)
                    
                    result.append(formatted)
                    i = j
                    continue

            result.append(words[i])
            i += 1

        return " ".join(result)

    @classmethod
    def normalize(cls, text: str) -> str:
        """Full pipeline for Vietnamese Inverse Text Normalization."""
        if not text or not text.strip():
            return text
        
        t = text
        t = cls.normalize_percentages(t)
        t = cls.normalize_dates_and_times(t)
        t = cls.normalize_phone_numbers(t)
        t = cls.normalize_general_numbers(t)
        
        # Clean up any duplicate spacing
        t = re.sub(r'[ \t]+', ' ', t).strip()
        return t

File: apps/desktop/src/utils/vietnamese-itn.ts
/**
 * Client-side Vietnamese Inverse Text Normalization (ITN).
 * Converts spoken number words to digits, percentages, and dates.
 * e.g.
 *  - "Ba ba bốn năm sáu Không ba bốn bốn" -> "33456 0344"
 *  - "hai mươi ba" -> "23"
 *  - "năm mươi lăm phần trăm" -> "55%"
 *  - "một triệu hai trăm nghìn đồng" -> "1.200.000 đồng"
 */

const UNITS: Record<string, number> = {
  không: 0,
  linh: 0,
  lẻ: 0,
  một: 1,
  mốt: 1,
  hai: 2,
  ba: 3,
  bốn: 4,
  tư: 4,
  năm: 5,
  lăm: 5,
  sáu: 6,
  bảy: 7,
  bẩy: 7,
  tám: 8,
  chín: 9,
};

const DIGIT_WORDS: Record<string, string> = {
  không: "0",
  một: "1",
  hai: "2",
  ba: "3",
  bốn: "4",
  tư: "4",
  năm: "5",
  sáu: "6",
  bảy: "7",
  bẩy: "7",
  tám: "8",
  chín: "9",
};

const NUMBER_KEYWORDS = new Set([
  ...Object.keys(UNITS),
  "mười",
  "mươi",
  "chục",
  "trăm",
  "nghìn",
  "ngàn",
  "triệu",
  "tỷ",
]);

export function parseVietnameseNumberWords(words: string[]): number | null {
  if (!words.length) return null;

  let currentBillions = 0;
  let currentMillions = 0;
  let currentThousands = 0;
  let currentHundreds = 0;
  let currentVal = 0;

  for (let i = 0; i < words.length; i++) {
    const w = words[i].toLowerCase();

    if (w === "tỷ") {
      const chunk =
        currentVal + currentHundreds + currentThousands + currentMillions || 1;
      currentBillions += chunk * 1_000_000_000;
      currentMillions = currentThousands = currentHundreds = currentVal = 0;
    } else if (w === "triệu") {
      const chunk = currentVal + currentHundreds + currentThousands || 1;
      currentMillions += chunk * 1_000_000;
      currentThousands = currentHundreds = currentVal = 0;
    } else if (w === "nghìn" || w === "ngàn") {
      const chunk = currentVal + currentHundreds || 1;
      currentThousands += chunk * 1_000;
      currentHundreds = currentVal = 0;
    } else if (w === "trăm") {
      const chunk = currentVal || 1;
      currentHundreds = chunk * 100;
      currentVal = 0;
    } else if (w === "mươi" || w === "chục") {
      const chunk = currentVal || 1;
      currentVal = chunk * 10;
    } else if (w === "mười") {
      currentVal += 10;
    } else if (w === "linh" || w === "lẻ") {
      // noop
    } else if (w in UNITS) {
      currentVal += UNITS[w];
    } else {
      return null;
    }
  }

  const total =
    currentBillions +
    currentMillions +
    currentThousands +
    currentHundreds +
    currentVal;
  return total;
}

export function formatNumberWithDots(n: number): string {
  if (n >= 10000) {
    return n.toLocaleString("vi-VN");
  }
  return String(n);
}

export function normalizeVietnameseNumbers(text: string): string {
  if (!text || !text.trim()) return text;

  let t = text;

  // 1. Percentages: "năm mươi lăm phần trăm" -> "55%"
  t = t.replace(
    /\b((?:(?:không|một|mốt|hai|ba|bốn|tư|năm|lăm|sáu|bảy|bẩy|tám|chín|mười|mươi|chục|trăm|linh|lẻ)\s+)+)phần trăm\b/gi,
    (match, numStr: string) => {
      const words = numStr.trim().split(/\s+/);
      const val = parseVietnameseNumberWords(words);
      return val !== null ? `${val}%` : match;
    },
  );
  t = t.replace(/(\d+)\s+phần trăm\b/gi, "$1%");

  // 2. Dates: "ngày mười lăm tháng tám năm hai nghìn..."
  t = t.replace(
    /\bngày\s+((?:không|mồng|mùng|một|mốt|hai|ba|bốn|tư|năm|lăm|sáu|bảy|bẩy|tám|chín|mười|mươi|chục|hai mươi|ba mươi)(?:\s+(?:một|mốt|hai|ba|bốn|tư|năm|lăm|sáu|bảy|bẩy|tám|chín))?)\b/gi,
    (match, dayStr: string) => {
      const words = dayStr.trim().split(/\s+/);
      const val = parseVietnameseNumberWords(words);
      return val !== null && val >= 1 && val <= 31 ? `ngày ${val}` : match;
    },
  );

  t = t.replace(
    /\btháng\s+((?:một|hai|ba|bốn|tư|năm|sáu|bảy|bẩy|tám|chín|mười|mười một|mười hai))\b/gi,
    (match, monthStr: string) => {
      const words = monthStr.trim().split(/\s+/);
      const val = parseVietnameseNumberWords(words);
      return val !== null && val >= 1 && val <= 12 ? `tháng ${val}` : match;
    },
  );

  t = t.replace(
    /\bnăm\s+((?:(?:một|hai)\s+nghìn(?:\s+(?:không|lẻ|linh|một|hai|ba|bốn|năm|sáu|bảy|tám|chín|mười|mươi|trăm))*))\b/gi,
    (match, yearStr: string) => {
      const words = yearStr.trim().split(/\s+/);
      const val = parseVietnameseNumberWords(words);
      return val !== null && val >= 1900 && val <= 2100 ? `năm ${val}` : match;
    },
  );

  // 3. Spoken phone numbers / digit sequences (>= 2 single digits): "ba ba bốn năm sáu" -> "33456"
  const tokens = t.split(/\s+/);
  const outTokens: string[] = [];
  let i = 0;
  while (i < tokens.length) {
    let j = i;
    const digits: string[] = [];
    while (j < tokens.length) {
      const cleanToken = tokens[j].toLowerCase().replace(/[^\w\s]/g, "");
      if (DIGIT_WORDS[cleanToken] !== undefined) {
        digits.push(DIGIT_WORDS[cleanToken]);
        j++;
      } else {
        break;
      }
    }
    if (digits.length >= 2) {
      let formattedDigits = digits.join("");
      const lastToken = tokens[j - 1];
      const punctMatch = lastToken.match(/([,.:;?!]+)$/);
      if (punctMatch) {
        formattedDigits += punctMatch[1];
      }
      outTokens.push(formattedDigits);
      i = j;
    } else {
      outTokens.push(tokens[i]);
      i++;
    }
  }
  t = outTokens.join(" ");

  // 4. General compound number spans (e.g. "hai mươi ba", "một trăm hai mươi nghìn")
  const words = t.split(/\s+/);
  const result: string[] = [];
  let wIdx = 0;
  while (wIdx < words.length) {
    const clean = words[wIdx].toLowerCase().replace(/[^\w\s]/g, "");
    if (NUMBER_KEYWORDS.has(clean)) {
      const span: string[] = [clean];
      const origSpan: string[] = [words[wIdx]];
      let nextIdx = wIdx + 1;
      while (nextIdx < words.length) {
        const nextClean = words[nextIdx].toLowerCase().replace(/[^\w\s]/g, "");
        if (NUMBER_KEYWORDS.has(nextClean)) {
          span.push(nextClean);
          origSpan.push(words[nextIdx]);
          nextIdx++;
        } else {
          break;
        }
      }

      if (span.length >= 2) {
        const val = parseVietnameseNumberWords(span);
        if (val !== null && val > 0) {
          let formatted = formatNumberWithDots(val);
          const lastOrig = origSpan[origSpan.length - 1];
          const punct = lastOrig.match(/([,.:;?!]+)$/);
          if (punct) {
            formatted += punct[1];
          }
          result.push(formatted);
          wIdx = nextIdx;
          continue;
        }
      }
    }
    result.push(words[wIdx]);
    wIdx++;
  }

  return result.join(" ");
}

/**
 * Chuẩn hóa khoảng trắng & định dạng khoảng cách dấu câu tự nhiên cho tiếng Việt.
 * Giữ nguyên dấu câu do mô hình (PhoVoice / Whisper) tự sinh, không tự ý chèn thêm dấu chấm câu.
 */
export function formatVietnamesePunctuation(text: string): string {
  if (!text) return "";

  // 1. Chuẩn hóa khoảng trắng & loại bỏ khoảng cách thừa
  let t = text.replace(/\s+/g, " ").trim();
  if (!t) return "";

  // 2. Chuẩn hóa khoảng cách xung quanh dấu câu nếu có sẵn từ mô hình
  // Xóa khoảng trắng trước dấu câu: "xin chào , bạn" -> "xin chào, bạn"
  t = t.replace(/\s+([.,!?:;…])/g, "$1");
  // Thêm khoảng trắng sau dấu câu nếu liền kề chữ cái: "chào.bạn" -> "chào. bạn"
  t = t.replace(/([.,!?:;…])(?=[^\s.,!?:;…\d])/g, "$1 ");

  // 3. Dọn dẹp các dấu câu bị lặp (ví dụ ".." -> ".", ",," -> ",")
  t = t.replace(/([.,!?:;])\1+/g, "$1");

  // 4. Sửa lỗi dấu chấm ngắt câu sai trước chữ thường trong tiếng Việt
  // (ví dụ: "thủy, băng, kim. hỏa mộc" -> "thủy, băng, kim, hỏa mộc")
  t = t.replace(
    /\.\s+(?=[a-zàáảãạăằắẳẵặâầấẩẫậèéẻẽẹêềếểễệìíỉĩịòóỏõọôồốổỗộơờớởỡợùúủũụưừứửữựỳýỷỹỵđ])/g,
    ", ",
  );

  // 5. Cụm từ liệt kê không thể kết thúc bằng dấu chấm
  t = t.replace(
    /(như là|bao gồm|gồm có|ví dụ như|cụ thể là|nghĩa là|tức là)\s*\.\s*$/gi,
    "$1,",
  );

  return t.trim();
}

/** Apply the same number, whitespace and punctuation cleanup everywhere a
 * transcript is persisted or rendered. Keeping this in one function prevents
 * realtime partials and final PhoVoice results from drifting apart. */
export function normalizeVietnameseTranscriptText(text: string): string {
  return formatVietnamesePunctuation(normalizeVietnameseNumbers(text));
}

/** Join transcript fragments without introducing spaces before punctuation or
 * preserving doubled whitespace from provider chunk boundaries. */
export function joinVietnameseTranscriptText(
  ...parts: Array<string | null | undefined>
): string {
  return normalizeVietnameseTranscriptText(
    parts.filter((part): part is string => Boolean(part?.trim())).join(" "),
  );
}

/** Remove text already present at the end of a committed transcript from a
 * provider's latest full hypothesis. Providers commonly resend the whole
 * sentence on every realtime update rather than a true delta. */
export function extractVietnameseTranscriptDelta(
  committedText: string,
  latestHypothesis: string,
): string {
  const committed = normalizeVietnameseTranscriptText(committedText);
  const latest = normalizeVietnameseTranscriptText(latestHypothesis);
  if (!latest || !committed) {
    return latest;
  }

  const canonicalize = (token: string) =>
    token.toLocaleLowerCase("vi").replace(/[.,!?;:…]+/g, "");
  const committedWords = committed
    .split(/\s+/)
    .map(canonicalize)
    .filter(Boolean);
  const latestTokens = latest.split(/\s+/);
  const latestWords = latestTokens.map(canonicalize).filter(Boolean);

  if (latestWords.join(" ") === committedWords.join(" ")) {
    return "";
  }

  const maxOverlap = Math.min(committedWords.length, latestWords.length);
  for (let overlap = maxOverlap; overlap >= 2; overlap--) {
    const committedSuffix = committedWords.slice(-overlap);
    const latestPrefix = latestWords.slice(0, overlap);
    if (committedSuffix.every((word, index) => word === latestPrefix[index])) {
      return normalizeVietnameseTranscriptText(
        latestTokens.slice(overlap).join(" "),
      );
    }
  }

  return latest;
}
